### DISCUSSION TITLE
Azure Service Bus migrate to premium - Connection rejected after GeoDR failover

### DISCUSSION THREAD

**Discussion by dldldlepl** [kedacore/keda#4686] [Q&A / Need Help]
### Report

When migrating a standard tier Azure Service Bus to premium tier the KEDA Scaler starts failing to connect with the error information: `Connection rejected after GeoDR failover`. Although the Connection String should be still valid as it becomes an alias connection string for the premium namespace.

### Expected Behavior

The KEDA Scaler has a downtime of about 5 minutes during the migration but is able to connect aterwards.

https://learn.microsoft.com/en-us/azure/service-bus-messaging/service-bus-migrate-standard-premium#how-much-downtime-do-i-expect

> How much downtime do I expect?
> ...
> The downtime that is experienced by the application is limited to the time it takes to update the DNS entry to point to the premium namespace. Downtime is approximately 5 minutes.

### Actual Behavior

After the migration from standard to premium tier is completed the KEDA Scaler logs events like:

```
GET https://XXX.servicebus.windows.net/TOPIC/Subscriptions/SUBSCRIPTION-------------------------------------------------------------------------------- RESPONSE 401: 401 Unauthorized ERROR CODE: 401 -------------------------------------------------------------------------------- Error Code 401 </Code> Detail Connection rejected after GeoDR failover. TrackingId:e6768e23-213f-476c-a654-f7bab349c5af_G4, SystemTracker:XXX.servicebus.windows.net:TOPIC/Subscriptions/SUBSCRIPTION, Timestamp:2023-05-16T13:24:18 </Detail> </Error> --------------------------------------------------------------------------------
```
and is not triggering new scaled jobs.

### Steps to Reproduce the Problem

1. Deploy a standard Azure Service Bus namespace
2. Create Topic
3. Create Subscription
4. Configure KEDA Scaler
```
triggers:
    - type: azure-servicebus
      metadata:
        topicName: TOPIC
        subscriptionName: SUBSCRIPTION
        messageCount: "1"
        connectionFromEnv: "SERVICEBUS_CONNECTIONSTRING"
        cloud: AzurePublicCloud
```
5. Send message into TOPIC
6. See that new jobs are started.
7. Consume message from SUBSCRIPTION
8. Migrate Service Bus from standard to premium
9. Wait about 5 minutes
10. Send message into TOPIC
11.  No new jobs are started and the KEDA scaler logs the shown error message.

### Logs from KEDA operator

```
GET https://XXX.servicebus.windows.net/TOPIC/Subscriptions/SUBSCRIPTION-------------------------------------------------------------------------------- RESPONSE 401: 401 Unauthorized ERROR CODE: 401 -------------------------------------------------------------------------------- Error Code 401 </Code> Detail Connection rejected after GeoDR failover. TrackingId:e6768e23-213f-476c-a654-f7bab349c5af_G4, SystemTracker:XXX.servicebus.windows.net:TOPIC/Subscriptions/SUBSCRIPTION, Timestamp:2023-05-16T13:24:18 </Detail> </Error> --------------------------------------------------------------------------------
```

### KEDA Version

2.9.3

### Kubernetes Version

1.24

### Platform

Microsoft Azure

### Scaler Details

Azure Service Bus

### Anything else?

Changing the configuration from e.g. the primary alias connection string to secondary alias connection string does not fix the problem. Changing from an alias connection string to the normal connection string of the premium namespace lets KEDA connect again. 

Switching back from the normal connection string to the alias connection string produces the error again.

Deleting the standard namespace does also fix this issue, but according to the documentation this only needs to be done to reduce later confusion. 

All other applications connected to the service bus instance did move over their connection as expected.

The issue was reproducible numerous times by migrating newly created an old existing standard namespaces to premium tier.

**Comment by JorTurFer** [comment]
@tomkerkhove . do you know if after migrating, anything else is required?

**Reply by tomkerkhove** [reply]
Are you sure this is related to the migration and not because of the geo failover that was triggered @dldldlepl?

**Reply by dldldlepl** [reply]
@tomkerkhove Yes this is related to the migration from standard to premium. A geo failover does not exist with standard namespaces as far as the documentation and configuration possibilities go because no pairing of namespaces can be activated in the standard tier. 

For the migration we used the following powershell script:
```
param (
    [Parameter(Mandatory=$true)]
    [string] $resourceGroup = "",
    [Parameter(Mandatory=$true)]
    [string] $standardNamespace = "",
    [Parameter(Mandatory=$true)]
    [string] $premiumNamespace = "",
    [Parameter(Mandatory=$true)]
    [string] $postMigrationDnsName = ""
)

$premiumNamespaceArmId = (Get-AzServiceBusNamespace -ResourceGroup $resourceGroup -Name $premiumNamespace).Id

Write-Host "ResourceGroup: $resourceGroup"
Write-Host "StandardNamespace: $standardNamespace"
Write-Host "PremiumNamespaceArmId: $premiumNamespaceArmId"
Write-Host "PostMigrationDnsName: $postMigrationDnsName"


Write-Host "Start migration to premimum namespace"

try {
    $entityMigrationStatus = Get-AzServiceBusMigration `
        -ResourceGroupName $resourceGroup `
        -NamespaceName $standardNamespace    
    
    Write-Host "Migration has been already executed"
    return;
}
catch {}


Start-AzServiceBusMigration `
    -ResourceGroupName $resourceGroup `
    -NamespaceName $standardNamespace `
    -TargetNamespace $premiumNamespaceArmId `
    -PostMigrationName $postMigrationDnsName

do 
{
    Write-Host "Wait for complete of migration to premimum namespace"

    $entityMigrationStatus = Get-AzServiceBusMigration `
        -ResourceGroupName $resourceGroup `
        -NamespaceName $standardNamespace

    Start-Sleep -Seconds 30
} while ($entityMigrationStatus.MigrationState -ne "Active" -and $entityMigrationStatus.PendingReplicationOperationsCount -ne 0 -and $entityMigrationStatus.ProvisioningState -ne "Succeeded")

Write-Host "Complete migration to premium namespace"

Complete-AzServiceBusMigration `
    -ResourceGroupName $resourceGroup `
    -NamespaceName $standardNamespace
```

**Reply by tomkerkhove** [reply]
Correct, but it is not because you have migrated that the issue is that, are you sure nobody triggered a geo failover? Because from the looks of it that is the issue, not the migration itself

**Reply by dldldlepl** [reply]
There is no paired premium namespace, so no failover can be triggered. And there was no trigger of a failover from our side. I don't know whether Azure can do something internally, but from our side no failover was triggered.

**Reply by tomkerkhove** [reply]
Can you open a support case with Microsoft to understand why you got that error please? I'm not sure if there is much we can do about that

**Reply by dldldlepl** [reply]
@tomkerkhove  I already opened a support ticket but until now they were still not very helpful, atleast they understand now that we have done the migration correctly and that it is not working as described in the documentation. But they want to close the ticket because they provided the workarround of deleting the standard namespace in the migration process which should not be necessary. (All other applications that we had did it correctly)

Some details I got through the support ticket so far are:
> 401 Unauthorized error is an [HTTP status code](https://www.lifewire.com/http-status-codes-2625907) that means the access token may be missing in the HTTP authenticate request header or that the token is invalid or has expired. Unfortunately, it’s not possible to check what exactly is not working, when KEDA gets this error messages.
> As we already informed you that after migration drain all the messages from the standard namespace and delete the standard namespace. If you not deleted, then the client would have open (TCP/TLS) connection to the standard namespace or sometime to premium namespace and therefore you get this error messages. So, you must have to delete the standard namespace after migration to premium.

So from that I suspect, that the problem is that the KEDA Scaler is not noticing, that it should open a new connection to the premium namespace. This would make sense as the connection string for the standard namespace changes. The DNS name changes to the post migration DNS name. A still open TCP/TLS connection to the standard namespace could then maybe cause that traffic in this connection is not valid as the SharedAccessSignature which is used for authentication has still the pre-migration DNS name and this would be invalid and therefore a 401 is returned.

**Reply by dldldlepl** [reply]
@JorTurFer, @tomkerkhove 
From the support ticket, that I opened: 
> As per my understanding you had raised the service request regarding the 401 unauthorized error you had encountered on migrating your Service Bus from Standard to Premium. While I do understand that the deletion of the Standard Namespace was conveyed to you as an important factor in the mitigation of this scenario. But, the main reason behind the 401 unauthorized in regards with the DNS mapping to the Old Standard Namespace and not the Premium Namespace. Post refreshing of the DNS records the connection would map to the Premium Namespace itself.

So it seems, that KEDA needs to do some DNS refreshing.

**Reply by JorTurFer** [reply]
In KEDA we don't have any kind of DNS, if it's something related with DNS, maybe restarting KEDA it solves the issue, but I'd point to the cluster DNS

**Reply by dldl-cmd** [reply]
I tried to debug this problem and I was able to reproduced it, by just follwing the Building & deploying locally steps and running against a locally running new minikube installation. From network tracing I was able to see, that the connection to the standard tier namespace is never terminated and requests are send to the standard namespace. No connection to the premium tier namespace is done.

As I have seen, that the implementation just uses the [azure-sdk-for-go](https://github.com/Azure/azure-sdk-for-go). I tried to implement a simple application and check its behavior. The following go application does also run into the same issue as described. 
```go
package main

import (
  "context"
  "fmt"
  "time"

  "github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus/admin"
)

func GetAdminClient() *admin.Client {
  connectionString := "<connection string>"
  client, err := admin.NewClientFromConnectionString(connectionString, nil)
  if err != nil {
    panic(err)
  }

  return client

}

func PrintMessageCountOrError(ctx context.Context, client *admin.Client) {
  properties, err := client.GetSubscriptionRuntimeProperties(ctx, "mytopic", "mysubscription", nil)
  if err != nil {
    fmt.Println(err)
  } else {
    fmt.Println(properties.TotalMessageCount)
  }
}

func main() {
  ctx := context.TODO()
  client := GetAdminClient()

  for {
    time.Sleep(1)
    PrintMessageCountOrError(ctx, client)
  }
}
```

**Reply by JorTurFer** [reply]
I have seen that you have already opened [an issue there](https://github.com/Azure/azure-sdk-for-go/issues/21172). Let's see if they solve the issue in the upstream 😄 
Just to understand the issue, does it happen only if KEDA is working and then, the servicebus tier is changed, right? Restarting KEDA pod is enough for fixing the issue because it's a DNS cache problem, am I right?

**Reply by dldl-cmd** [reply]
Yes, in normal operation there is no problem because the IP address of the ServiceBus seems to be kept stable: https://learn.microsoft.com/en-us/azure/service-bus-messaging/service-bus-faq#what-ip-addresses-do-i-need-to-add-to-allowlist-

There is mentioned that there are reasons why the IP-Address might change, which is also not complete, because the migration does change the IP address:
> The IP address returned by the nslookup command isn't a static IP address. However, it remains constant until the underlying deployment is deleted or moved to a different cluster.

Furthermore, an infrastructure update like announced [here](https://techcommunity.microsoft.com/t5/messaging-on-azure-blog/upcoming-changes-to-ip-addresses-for-azure-service-bus/ba-p/3285247) might also change the IP address.

From what I understand the problem could always occur, when the IP address changes and the connection to the old IP address is still functioning. Therefore, deleting the standard namespace helps because the connection will get closed, or staying longer in the debugger which will cause a timeout on the connection fixes the problem. Also restarting KEDA fixes the problem.

**Reply by JorTurFer** [reply]
Thanks for sharing the info!
I have subscribed the upstream issue, once it's solved there, we will bump the version to address it in KEDA ❤️
